{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create a table with SRA IDs, readstats QC data, and submitter metadata\n",
    "\n",
    "**Below are the steps taken in this notebook:**\n",
    "1. Installs, import Statements & Global Variable Definitions\n",
    "2. Read in tables as dataframes\n",
    "3. Merge dataframes\n",
    "4. Write merged dataframe to data table and file\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "%pip install gcsfs\n",
    "## capture CANNOT have comments above it\n",
    "## For reading CSVs stored in Google Cloud (without downloading them first)\n",
    "## May need to restart kernel after install "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "%pip install --upgrade --no-cache-dir --force-reinstall terra-pandas\n",
    "%pip install --upgrade --no-cache-dir  --force-reinstall git+https://github.com/DataBiosphere/terra-notebook-utils\n",
    "## For reading/writing data tables into pandas data frames\n",
    "## May need to restart kernel after install "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Statements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from firecloud import fiss\n",
    "import pandas as pd      \n",
    "import os                 \n",
    "import subprocess       \n",
    "import re                 \n",
    "import io\n",
    "import gcsfs\n",
    "\n",
    "from typing import Any, Callable, List, Optional\n",
    "from terra_notebook_utils import table, WORKSPACE_NAME, WORKSPACE_GOOGLE_PROJECT\n",
    "from terra_pandas import dataframe_to_table, table_to_dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Global Variable Declarations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Billing project: human-pangenome-ucsc\n",
      "Workspace: HPRC_WRANGLING_UW_HPRC_HiFi_Y1\n",
      "Workspace storage bucket: gs://fc-ed02327c-c4c0-4a7a-8339-50e7ebe38ce5/\n"
     ]
    }
   ],
   "source": [
    "# AnVIL_HPRC WorkspaceBucket\n",
    "anvil_hprc_bucket       = \"gs://fc-4310e737-a388-4a10-8c9e-babe06aaf0cf/\"\n",
    "\n",
    "# table filenames (expected in the workspace bucket, see below)\n",
    "submitter_metadata = '201112_UW_HPRC_PacBio_HiFi_Metadata_Submission_v0.2_kmmod.txt'\n",
    "sra_metadata = 'metadata-8323615-processed-ok.tsv'\n",
    "\n",
    "# submission id\n",
    "submission_id = 'UW_HPRC_HiFi_Y1'\n",
    "\n",
    "# Get the Google billing project name and workspace name for current workspace\n",
    "PROJECT = os.environ['WORKSPACE_NAMESPACE']\n",
    "WORKSPACE =os.path.basename(os.path.dirname(os.getcwd()))\n",
    "bucket = os.environ['WORKSPACE_BUCKET'] + \"/\"\n",
    "\n",
    "# Verify that we've captured the environment variables\n",
    "print(\"Billing project: \" + PROJECT)\n",
    "print(\"Workspace: \" + WORKSPACE)\n",
    "print(\"Workspace storage bucket: \" + bucket)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read In tables as dataframes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Readstats table\n",
    "Find the individual readstats output files in the table and concatenate their contents.\n",
    "Add the hifi filename and sample name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1</th>\n",
       "      <th>total_reads</th>\n",
       "      <th>total_bp</th>\n",
       "      <th>total_Gbp</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>mean</th>\n",
       "      <th>quartile_25</th>\n",
       "      <th>quartile_50</th>\n",
       "      <th>quartile_75</th>\n",
       "      <th>N25</th>\n",
       "      <th>N50</th>\n",
       "      <th>N75</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1238949</td>\n",
       "      <td>22838743498</td>\n",
       "      <td>22.84</td>\n",
       "      <td>48</td>\n",
       "      <td>43615</td>\n",
       "      <td>18433</td>\n",
       "      <td>17496</td>\n",
       "      <td>18376</td>\n",
       "      <td>19340</td>\n",
       "      <td>17588</td>\n",
       "      <td>18478</td>\n",
       "      <td>19450</td>\n",
       "      <td>m54329U_201103_231616.ccs.bam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>849857</td>\n",
       "      <td>17328670239</td>\n",
       "      <td>17.33</td>\n",
       "      <td>46</td>\n",
       "      <td>57023</td>\n",
       "      <td>20390</td>\n",
       "      <td>18712</td>\n",
       "      <td>19835</td>\n",
       "      <td>21255</td>\n",
       "      <td>18875</td>\n",
       "      <td>20054</td>\n",
       "      <td>21609</td>\n",
       "      <td>m64076_201013_225902.ccs.bam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1462110</td>\n",
       "      <td>29289007356</td>\n",
       "      <td>29.29</td>\n",
       "      <td>45</td>\n",
       "      <td>50257</td>\n",
       "      <td>20032</td>\n",
       "      <td>17344</td>\n",
       "      <td>19636</td>\n",
       "      <td>22397</td>\n",
       "      <td>17945</td>\n",
       "      <td>20385</td>\n",
       "      <td>23250</td>\n",
       "      <td>m64076_200311_082315.ccs.bam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1610516</td>\n",
       "      <td>30068543676</td>\n",
       "      <td>30.07</td>\n",
       "      <td>47</td>\n",
       "      <td>50034</td>\n",
       "      <td>18670</td>\n",
       "      <td>16676</td>\n",
       "      <td>18344</td>\n",
       "      <td>20318</td>\n",
       "      <td>17038</td>\n",
       "      <td>18766</td>\n",
       "      <td>20859</td>\n",
       "      <td>m64076_200201_051547.ccs.bam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1779397</td>\n",
       "      <td>32630636515</td>\n",
       "      <td>32.63</td>\n",
       "      <td>44</td>\n",
       "      <td>49139</td>\n",
       "      <td>18338</td>\n",
       "      <td>16407</td>\n",
       "      <td>18053</td>\n",
       "      <td>19983</td>\n",
       "      <td>16771</td>\n",
       "      <td>18467</td>\n",
       "      <td>20512</td>\n",
       "      <td>m64076_200203_181219.ccs.bam</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "1 total_reads     total_bp total_Gbp min    max   mean quartile_25  \\\n",
       "2     1238949  22838743498     22.84  48  43615  18433       17496   \n",
       "2      849857  17328670239     17.33  46  57023  20390       18712   \n",
       "2     1462110  29289007356     29.29  45  50257  20032       17344   \n",
       "2     1610516  30068543676     30.07  47  50034  18670       16676   \n",
       "2     1779397  32630636515     32.63  44  49139  18338       16407   \n",
       "\n",
       "1 quartile_50 quartile_75    N25    N50    N75                       filename  \n",
       "2       18376       19340  17588  18478  19450  m54329U_201103_231616.ccs.bam  \n",
       "2       19835       21255  18875  20054  21609   m64076_201013_225902.ccs.bam  \n",
       "2       19636       22397  17945  20385  23250   m64076_200311_082315.ccs.bam  \n",
       "2       18344       20318  17038  18766  20859   m64076_200201_051547.ccs.bam  \n",
       "2       18053       19983  16771  18467  20512   m64076_200203_181219.ccs.bam  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# readstats\n",
    "readstats_df = table_to_dataframe(\"readstats\", workspace=WORKSPACE, workspace_namespace=PROJECT)\n",
    "qc_list = []\n",
    "\n",
    "for index, row in readstats_df.iterrows():\n",
    "\n",
    "        sample_readstats_fp = row['ReadStatsReport']\n",
    "        sample_readstats_fn = os.path.basename(sample_readstats_fp)\n",
    "        df = pd.read_csv(sample_readstats_fp, header=None, sep='\\t')\n",
    "        df = df[df[0]=='sample.fastq']\n",
    "        # add sample and file name\n",
    "        #df = df.append({0:'undef', 1:'sample', 2:row['sample']}, ignore_index=True)\n",
    "        df = df.append({0:'undef', 1:'filename', 2:os.path.basename(row['hifi'])}, ignore_index=True)\n",
    "        # remove all but the variables (total_bp, quartile_25 etc)\n",
    "        df2 = df.drop(df.columns[[0, 1]], axis=1)\n",
    "        # make rownames ('total_bp', 'quartile_25' etc)\n",
    "        df2.index = df.iloc[:,1]\n",
    "        # remove the now meaningless 'sample.fastq' filename\n",
    "        df2 = df2.drop(index=['file'])\n",
    "        qc_list.append(df2)\n",
    "\n",
    "# merge\n",
    "readstats_df = pd.concat(qc_list, axis=1).transpose()\n",
    "readstats_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submitter metadata table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>shear_method</th>\n",
       "      <th>size_selection</th>\n",
       "      <th>ccs_algorithm</th>\n",
       "      <th>polymerase_version</th>\n",
       "      <th>seq_plate_chemistry_version</th>\n",
       "      <th>generator_facility</th>\n",
       "      <th>generator_contact</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>m54329U_200201_051510.ccs.bam</td>\n",
       "      <td>HG01123</td>\n",
       "      <td>PacBio Sequel II</td>\n",
       "      <td>g-TUBE</td>\n",
       "      <td>SageELF</td>\n",
       "      <td>ccs 4.0.0 (commit SL-release-8.0.0)</td>\n",
       "      <td>P2.0</td>\n",
       "      <td>C2.0</td>\n",
       "      <td>University of Washington</td>\n",
       "      <td>kmiyamot@uw.edu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>m54329U_200203_181235.ccs.bam</td>\n",
       "      <td>HG01123</td>\n",
       "      <td>PacBio Sequel II</td>\n",
       "      <td>g-TUBE</td>\n",
       "      <td>SageELF</td>\n",
       "      <td>ccs 4.0.0 (commit SL-release-8.0.0)</td>\n",
       "      <td>P2.0</td>\n",
       "      <td>C2.0</td>\n",
       "      <td>University of Washington</td>\n",
       "      <td>kmiyamot@uw.edu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>m54329U_200205_002609.ccs.bam</td>\n",
       "      <td>HG01123</td>\n",
       "      <td>PacBio Sequel II</td>\n",
       "      <td>g-TUBE</td>\n",
       "      <td>SageELF</td>\n",
       "      <td>ccs 4.0.0 (commit SL-release-8.0.0)</td>\n",
       "      <td>P2.0</td>\n",
       "      <td>C2.0</td>\n",
       "      <td>University of Washington</td>\n",
       "      <td>kmiyamot@uw.edu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>m54329U_200206_215926.ccs.bam</td>\n",
       "      <td>HG01123</td>\n",
       "      <td>PacBio Sequel II</td>\n",
       "      <td>g-TUBE</td>\n",
       "      <td>SageELF</td>\n",
       "      <td>ccs 4.0.0 (commit SL-release-8.0.0)</td>\n",
       "      <td>P2.0</td>\n",
       "      <td>C2.0</td>\n",
       "      <td>University of Washington</td>\n",
       "      <td>kmiyamot@uw.edu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>m64076_200306_185917.ccs.bam</td>\n",
       "      <td>HG01258</td>\n",
       "      <td>PacBio Sequel II</td>\n",
       "      <td>g-TUBE</td>\n",
       "      <td>SageELF</td>\n",
       "      <td>ccs 4.0.0 (commit SL-release-8.0.0)</td>\n",
       "      <td>P2.0</td>\n",
       "      <td>C2.0</td>\n",
       "      <td>University of Washington</td>\n",
       "      <td>kmiyamot@uw.edu</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        filename sample_ID  instrument_model shear_method  \\\n",
       "0  m54329U_200201_051510.ccs.bam   HG01123  PacBio Sequel II       g-TUBE   \n",
       "1  m54329U_200203_181235.ccs.bam   HG01123  PacBio Sequel II       g-TUBE   \n",
       "2  m54329U_200205_002609.ccs.bam   HG01123  PacBio Sequel II       g-TUBE   \n",
       "3  m54329U_200206_215926.ccs.bam   HG01123  PacBio Sequel II       g-TUBE   \n",
       "4   m64076_200306_185917.ccs.bam   HG01258  PacBio Sequel II       g-TUBE   \n",
       "\n",
       "  size_selection                        ccs_algorithm polymerase_version  \\\n",
       "0        SageELF  ccs 4.0.0 (commit SL-release-8.0.0)               P2.0   \n",
       "1        SageELF  ccs 4.0.0 (commit SL-release-8.0.0)               P2.0   \n",
       "2        SageELF  ccs 4.0.0 (commit SL-release-8.0.0)               P2.0   \n",
       "3        SageELF  ccs 4.0.0 (commit SL-release-8.0.0)               P2.0   \n",
       "4        SageELF  ccs 4.0.0 (commit SL-release-8.0.0)               P2.0   \n",
       "\n",
       "  seq_plate_chemistry_version        generator_facility generator_contact  \n",
       "0                        C2.0  University of Washington   kmiyamot@uw.edu  \n",
       "1                        C2.0  University of Washington   kmiyamot@uw.edu  \n",
       "2                        C2.0  University of Washington   kmiyamot@uw.edu  \n",
       "3                        C2.0  University of Washington   kmiyamot@uw.edu  \n",
       "4                        C2.0  University of Washington   kmiyamot@uw.edu  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_fp = os.path.join(bucket, submitter_metadata)\n",
    "metadata_df = pd.read_csv(meta_fp, sep='\\t')\n",
    "metadata_df = metadata_df[['filename', 'sample_ID', 'instrument_model', 'shear_method', 'size_selection', 'ccs_algorithm', \n",
    "    'polymerase_version', 'seq_plate_chemistry_version', 'generator_facility', 'generator_contact']]\n",
    "metadata_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SRA table\n",
    "Must download after finishing submission.\n",
    "The downloaded table needs to be split in one row per file (instead of filename, filename2 <...> filenameN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>accession</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>m54329U_200201_051510.ccs.bam</td>\n",
       "      <td>SRR13684290</td>\n",
       "      <td>HG01123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>m64076_200306_185917.ccs.bam</td>\n",
       "      <td>SRR13684289</td>\n",
       "      <td>HG01258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>m64076_200201_051547.ccs.bam</td>\n",
       "      <td>SRR13684283</td>\n",
       "      <td>HG01358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>m54329U_200306_185930.ccs.bam</td>\n",
       "      <td>SRR13684282</td>\n",
       "      <td>HG01361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>m64076_200221_234426.ccs.bam</td>\n",
       "      <td>SRR13684281</td>\n",
       "      <td>HG01361</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        filename    accession   sample\n",
       "0  m54329U_200201_051510.ccs.bam  SRR13684290  HG01123\n",
       "1   m64076_200306_185917.ccs.bam  SRR13684289  HG01258\n",
       "2   m64076_200201_051547.ccs.bam  SRR13684283  HG01358\n",
       "3  m54329U_200306_185930.ccs.bam  SRR13684282  HG01361\n",
       "4   m64076_200221_234426.ccs.bam  SRR13684281  HG01361"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Read in SRA file and split by filename\n",
    "\n",
    "sra_file = os.path.join(bucket + sra_metadata)\n",
    "sample_df = pd.read_csv(sra_file, sep='\\t')\n",
    "\n",
    "fnames = [x for x in sample_df.columns if x.startswith('filename')]\n",
    "sra_df = pd.melt(sample_df, id_vars =['sample_name', 'accession'], value_vars = fnames, value_name='file')\n",
    "sra_df= sra_df.dropna()\n",
    "sra_df = sra_df[['file', 'accession', 'sample_name']]\n",
    "sra_df.columns = ['filename', 'accession', 'sample']\n",
    "sra_df.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge all dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(46, 3) (47, 13) (46, 10)\n"
     ]
    }
   ],
   "source": [
    "print(sra_df.shape, readstats_df.shape, metadata_df.shape)\n",
    "# two samples of HG02572 were not uploaded to SRA with the rest of this batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'m54329U_201103_231616.ccs.bam'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(readstats_df['filename']) - set(sra_df['filename'])\n",
    "#m54329U_201103_231616.ccs.bam (HG002 sample) is not present in the submitter metadata or on SRA, but it is in the submissions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>accession</th>\n",
       "      <th>sample</th>\n",
       "      <th>total_reads</th>\n",
       "      <th>total_bp</th>\n",
       "      <th>total_Gbp</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>mean</th>\n",
       "      <th>quartile_25</th>\n",
       "      <th>...</th>\n",
       "      <th>N75</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>shear_method</th>\n",
       "      <th>size_selection</th>\n",
       "      <th>ccs_algorithm</th>\n",
       "      <th>polymerase_version</th>\n",
       "      <th>seq_plate_chemistry_version</th>\n",
       "      <th>generator_facility</th>\n",
       "      <th>generator_contact</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>m54329U_200201_051510.ccs.bam</td>\n",
       "      <td>SRR13684290</td>\n",
       "      <td>HG01123</td>\n",
       "      <td>2072143</td>\n",
       "      <td>34554070679</td>\n",
       "      <td>34.55</td>\n",
       "      <td>46</td>\n",
       "      <td>43126</td>\n",
       "      <td>16675</td>\n",
       "      <td>14984</td>\n",
       "      <td>...</td>\n",
       "      <td>18566</td>\n",
       "      <td>HG01123</td>\n",
       "      <td>PacBio Sequel II</td>\n",
       "      <td>g-TUBE</td>\n",
       "      <td>SageELF</td>\n",
       "      <td>ccs 4.0.0 (commit SL-release-8.0.0)</td>\n",
       "      <td>P2.0</td>\n",
       "      <td>C2.0</td>\n",
       "      <td>University of Washington</td>\n",
       "      <td>kmiyamot@uw.edu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>m64076_200306_185917.ccs.bam</td>\n",
       "      <td>SRR13684289</td>\n",
       "      <td>HG01258</td>\n",
       "      <td>1353905</td>\n",
       "      <td>26041896578</td>\n",
       "      <td>26.04</td>\n",
       "      <td>44</td>\n",
       "      <td>49803</td>\n",
       "      <td>19234</td>\n",
       "      <td>16739</td>\n",
       "      <td>...</td>\n",
       "      <td>22588</td>\n",
       "      <td>HG01258</td>\n",
       "      <td>PacBio Sequel II</td>\n",
       "      <td>g-TUBE</td>\n",
       "      <td>SageELF</td>\n",
       "      <td>ccs 4.0.0 (commit SL-release-8.0.0)</td>\n",
       "      <td>P2.0</td>\n",
       "      <td>C2.0</td>\n",
       "      <td>University of Washington</td>\n",
       "      <td>kmiyamot@uw.edu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>m64076_200201_051547.ccs.bam</td>\n",
       "      <td>SRR13684283</td>\n",
       "      <td>HG01358</td>\n",
       "      <td>1610516</td>\n",
       "      <td>30068543676</td>\n",
       "      <td>30.07</td>\n",
       "      <td>47</td>\n",
       "      <td>50034</td>\n",
       "      <td>18670</td>\n",
       "      <td>16676</td>\n",
       "      <td>...</td>\n",
       "      <td>20859</td>\n",
       "      <td>HG01358</td>\n",
       "      <td>PacBio Sequel II</td>\n",
       "      <td>g-TUBE</td>\n",
       "      <td>SageELF</td>\n",
       "      <td>ccs 4.0.0 (commit SL-release-8.0.0)</td>\n",
       "      <td>P2.0</td>\n",
       "      <td>C2.0</td>\n",
       "      <td>University of Washington</td>\n",
       "      <td>kmiyamot@uw.edu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>m54329U_200306_185930.ccs.bam</td>\n",
       "      <td>SRR13684282</td>\n",
       "      <td>HG01361</td>\n",
       "      <td>1662447</td>\n",
       "      <td>33289500589</td>\n",
       "      <td>33.29</td>\n",
       "      <td>46</td>\n",
       "      <td>49882</td>\n",
       "      <td>20024</td>\n",
       "      <td>17940</td>\n",
       "      <td>...</td>\n",
       "      <td>22241</td>\n",
       "      <td>HG01361</td>\n",
       "      <td>PacBio Sequel II</td>\n",
       "      <td>g-TUBE</td>\n",
       "      <td>SageELF</td>\n",
       "      <td>ccs 4.0.0 (commit SL-release-8.0.0)</td>\n",
       "      <td>P2.0</td>\n",
       "      <td>C2.0</td>\n",
       "      <td>University of Washington</td>\n",
       "      <td>kmiyamot@uw.edu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>m64076_200221_234426.ccs.bam</td>\n",
       "      <td>SRR13684281</td>\n",
       "      <td>HG01361</td>\n",
       "      <td>352516</td>\n",
       "      <td>6760635369</td>\n",
       "      <td>6.76</td>\n",
       "      <td>45</td>\n",
       "      <td>42847</td>\n",
       "      <td>19178</td>\n",
       "      <td>18750</td>\n",
       "      <td>...</td>\n",
       "      <td>20718</td>\n",
       "      <td>HG01361</td>\n",
       "      <td>PacBio Sequel II</td>\n",
       "      <td>g-TUBE</td>\n",
       "      <td>SageELF</td>\n",
       "      <td>ccs 4.0.0 (commit SL-release-8.0.0)</td>\n",
       "      <td>P2.0</td>\n",
       "      <td>C2.0</td>\n",
       "      <td>University of Washington</td>\n",
       "      <td>kmiyamot@uw.edu</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        filename    accession   sample total_reads  \\\n",
       "0  m54329U_200201_051510.ccs.bam  SRR13684290  HG01123     2072143   \n",
       "1   m64076_200306_185917.ccs.bam  SRR13684289  HG01258     1353905   \n",
       "2   m64076_200201_051547.ccs.bam  SRR13684283  HG01358     1610516   \n",
       "3  m54329U_200306_185930.ccs.bam  SRR13684282  HG01361     1662447   \n",
       "4   m64076_200221_234426.ccs.bam  SRR13684281  HG01361      352516   \n",
       "\n",
       "      total_bp total_Gbp min    max   mean quartile_25  ...    N75 sample_ID  \\\n",
       "0  34554070679     34.55  46  43126  16675       14984  ...  18566   HG01123   \n",
       "1  26041896578     26.04  44  49803  19234       16739  ...  22588   HG01258   \n",
       "2  30068543676     30.07  47  50034  18670       16676  ...  20859   HG01358   \n",
       "3  33289500589     33.29  46  49882  20024       17940  ...  22241   HG01361   \n",
       "4   6760635369      6.76  45  42847  19178       18750  ...  20718   HG01361   \n",
       "\n",
       "   instrument_model shear_method size_selection  \\\n",
       "0  PacBio Sequel II       g-TUBE        SageELF   \n",
       "1  PacBio Sequel II       g-TUBE        SageELF   \n",
       "2  PacBio Sequel II       g-TUBE        SageELF   \n",
       "3  PacBio Sequel II       g-TUBE        SageELF   \n",
       "4  PacBio Sequel II       g-TUBE        SageELF   \n",
       "\n",
       "                         ccs_algorithm polymerase_version  \\\n",
       "0  ccs 4.0.0 (commit SL-release-8.0.0)               P2.0   \n",
       "1  ccs 4.0.0 (commit SL-release-8.0.0)               P2.0   \n",
       "2  ccs 4.0.0 (commit SL-release-8.0.0)               P2.0   \n",
       "3  ccs 4.0.0 (commit SL-release-8.0.0)               P2.0   \n",
       "4  ccs 4.0.0 (commit SL-release-8.0.0)               P2.0   \n",
       "\n",
       "  seq_plate_chemistry_version        generator_facility generator_contact  \n",
       "0                        C2.0  University of Washington   kmiyamot@uw.edu  \n",
       "1                        C2.0  University of Washington   kmiyamot@uw.edu  \n",
       "2                        C2.0  University of Washington   kmiyamot@uw.edu  \n",
       "3                        C2.0  University of Washington   kmiyamot@uw.edu  \n",
       "4                        C2.0  University of Washington   kmiyamot@uw.edu  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df = pd.merge(\n",
    "    sra_df,\n",
    "    readstats_df,\n",
    "    on='filename')\n",
    "merged_df = pd.merge(\n",
    "    merged_df,\n",
    "    metadata_df,\n",
    "    on='filename')\n",
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload To Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create table\n",
    "#dataframe_to_table(\"post_sra_metadata\", merged_df, WORKSPACE, PROJECT)\n",
    "outf = os.path.join(bucket, submission_id + '_post_sra_metadata.tsv')\n",
    "merged_df.to_csv(outf, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46, 24)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
